### DISCUSSION TITLE
how i can use a cosine index

### DISCUSSION THREAD

**Discussion by othmanjad** [milvus-io/milvus#26892] [Q&A and General discussion]
hello every one 

i have a this code example 

`

from pymilvus import connections, FieldSchema, CollectionSchema, DataType, Collection, utility
import pandas as pd
import numpy as np
df = pd.read_csv('New_Medium_Data.csv', converters={'title_vector': lambda x: eval(x)})
df.head()
connections.connect(host='localhost', port='19530')

def create_milvus_collection(collection_name, dim):
    if utility.has_collection(collection_name):
        utility.drop_collection(collection_name)
    
    fields = [
            FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
            FieldSchema(name="title", dtype=DataType.VARCHAR, max_length=500),   
            FieldSchema(name="title_vector", dtype=DataType.FLOAT_VECTOR, dim=dim),
            FieldSchema(name="link", dtype=DataType.VARCHAR, max_length=500),
            FieldSchema(name="reading_time", dtype=DataType.INT64),
            FieldSchema(name="publication", dtype=DataType.VARCHAR, max_length=500),
            FieldSchema(name="claps", dtype=DataType.INT64),
            FieldSchema(name="responses", dtype=DataType.INT64)
    ]
    schema = CollectionSchema(fields=fields, description='search text')
    collection = Collection(name=collection_name, schema=schema)
    
    index_params = {
        'metric_type': "L2",
        'index_type': "IVF_FLAT",
        'params': {"nlist": 2048}
    }
    collection.create_index(field_name='title_vector', index_params=index_params)
    return collection

collection = create_milvus_collection('search_article_in_medium', 768)

from towhee import ops, pipe, DataCollection

insert_pipe = (pipe.input('df')
                   .flat_map('df', 'data', lambda df: df.values.tolist())
                   .map('data', 'res', ops.ann_insert.milvus_client(host='localhost', 
                                                                    port='19530',
                                                                    collection_name='search_article_in_medium'))
                   .output('res')
)
insert_pipe(df)

collection.load()
collection.num_entities

search_pipe1 = (pipe.input('query')
                    .map('query', 'vec', ops.text_embedding.dpr(model_name="facebook/dpr-ctx_encoder-single-nq-base"))
                    .map('vec', 'vec', lambda x: x / np.linalg.norm(x, axis=0))
                    .flat_map('vec', ('id', 'score', 'title'), ops.ann_search.milvus_client(host='127.0.0.1', 
                                                                                   port='19530',
                                                                                   collection_name='search_article_in_medium',
                                                                                   output_fields=['title']))  
                    .output('query', 'id', 'score', 'title')
               )

res = search_pipe1('Adventure into Machine Learning using Python')
DataCollection(res).show()

`

form this link 
https://github.com/towhee-io/examples/blob/main/nlp/text_search/search_article_in_medium.ipynb

can any one help me to convert it from l2 to cosine index

**Comment by xiaofan-luan** [comment]
you just need to change
metric_type from L2 to cosine

**Comment by yhmo** [comment]
The Milvus v2.3.0 supports COSINE metric:
```
index_params = {
    'metric_type': "COSINE",
    'index_type': "IVF_FLAT",
    'params': {"nlist": 2048}
}
collection.create_index(field_name='title_vector', index_params=index_params)
```

**Reply by othmanjad** [reply]
Why a search data not  is inaccurate

**Reply by xiaofan-luan** [reply]
uh? what do you mean by "not is inaccurate"
